src/main/java/bio/terra/folder/db/FolderDao.java

```diff
@@ -0,0 +1,102 @@
+
+  @Autowired
+  public FolderDao(NamedParameterJdbcTemplate jdbcTemplate) {
+    this.jdbcTemplate = jdbcTemplate;
+  }
+
+  public void createFolder(
+      String folderId,
+      String folderName,
+      JsonNullable<String> parentFolderId,
+      JsonNullable<String> spendProfile,
+      boolean spendProfileInherited) {
+    String sql =
+        "INSERT INTO folder (folder_id, folder_name, parent_folder_id, spend_profile_id, spend_profile_inherited)"
+            + "values (:folder_id, :folder_name, :parent_folder_id, :spend_profile_id, :spend_profile_inherited)";
+    Map<String, Object> paramMap = new HashMap<>();
+    paramMap.put("folder_id", folderId);
+    paramMap.put("folder_name", folderName);
+    paramMap.put("spend_profile_inherited", spendProfileInherited);
+
+    if (parentFolderId.isPresent() && parentFolderId.get() != null) {
+      paramMap.put("parent_folder_id", parentFolderId.get());
+    } else {
+      paramMap.put("parent_folder_id", null);
+    }
+
+    if (spendProfile.isPresent() && spendProfile.get() != null) {
+      paramMap.put("spend_profile_id", spendProfile.get());
+    } else {
+      paramMap.put("spend_profile_id", null);
+    }
+
+    jdbcTemplate.update(sql, paramMap);
+  }
+
+  public boolean deleteFolder(String folderId) {
+    String sql = "DELETE FROM folder WHERE folder_id = :id";
+    Map<String, Object> paramMap = new HashMap<>();
+    paramMap.put("id", folderId);
+    int rowsAffected = jdbcTemplate.update(sql, paramMap);
+    return rowsAffected > 0;
+  }
+
+  public String getSpendProfileFromFolder(String folderId) {
+    String sql = "SELECT spend_profile_id from folder WHERE folder_id = :id";
+    Map<String, Object> paramMap = new HashMap<>();
+    paramMap.put("id", folderId);
+    try {
+      return jdbcTemplate.queryForObject(sql, paramMap, String.class);
+    } catch (EmptyResultDataAccessException e) {
+      throw new FolderNotFoundException("Folder not found in DB: " + folderId);
+    }
+  }
+
+  public boolean getSpendProfileInheritedFromFolder(String folderId) {
+    String sql = "SELECT spend_profile_inherited from folder WHERE folder_id = :id";
+    Map<String, Object> paramMap = new HashMap<>();
+    paramMap.put("id", folderId);
+    try {
+      return jdbcTemplate.queryForObject(sql, paramMap, Boolean.class);
+    } catch (EmptyResultDataAccessException e) {
+      throw new FolderNotFoundException("Folder not found in DB: " + folderId);
+    }
+  }
+
+  // Checks whether a given name is taken by a contained object within a folder.
+  public boolean containedObjectNameExists(String parentFolderId, String name) {
+    String sql = "SELECT * from contained_object WHERE folder_id = :id AND object_name = :name";
+    Map<String, Object> paramMap = new HashMap<>();
+    paramMap.put("id", parentFolderId);
+    paramMap.put("name", name);
+
+    List<Map<String, Object>> resultSet = jdbcTemplate.queryForList(sql, paramMap);
+    return !resultSet.isEmpty();
+  }
```

Comment:
could you optimize the SQL for this?  I realize that business logic says it should always return 0 or 1 rows, but it would be good to encode that into the implementation. `SELECT TOP 1 object_id from ... LIMIT 1` is an option, or `SELECT COUNT(1) from ... ` or using the `EXISTS` operator.

( apply to `containedFolderNameExists` also )

Reply:
Went with LIMIT 1, I don't think postgres has a TOP operator (but I might be missing it)